### PR TITLE
Introduce "base" and "root" token definitions

### PR DESCRIPTION
**Pull request by Padraic-O-Mhuiris**
Currently this pr is targeting the [add array extensions](https://github.com/Uniswap/token-lists.git#add-array-extension-support) pr in the tokenlists repo which as stated adds array types to the extensions object. This is important to us because in our system, a majority of our current principal tokens are dependent on "underlying" curve pool "LP" tokens, which themselves are representative of a number of "vanilla" tokens which are exchanged on the pool. This PR includes the renaming of the "underlying" token to a "base" token as I interpret our terms to be "based" on these, this is possibly breaking but I think it is preferable. The "vanilla" tokens as described are introduced and named as "root" tokens definitions.
The rationale behind the inclusion of the array extensions is that the nature of the curve pool contracts requires us to interface with them using fixed-length address arrays which for each index, a correspondence exists to a "root" token in the pool. E.g, the  `LUSD-3CRV-f" base token's pool has defined two tokens, `LUSD` and `3CRV`. We define in the "base" token definition an array of "poolAssets" which is a string array of length 2 or 3 which would contain the addresses for these tokens.

We may want to wait until the array extension PR is merged as I have used the github branch url in the absence of an npm package

### CODE REVIEW

File: src/curveToken.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  const curvePoolContract = new ethers.Contract(pool, curvePoolAbi, provider);
+
+  const addLiquidityFuncSig = Object.keys(curvePoolContract.functions).find(
+    (k) => k.startsWith("add_liquidity")
+  ) as string;
+  const removeLiquidityFuncSig = Object.keys(curvePoolContract.functions).find(
+    (k) => k.startsWith("remove_liquidity_one_coin")
+  ) as string;
+
+  const numCoins = parseInt(
```
Comment (sentilesdal) [inline src/curveToken.ts:173]: add comment with function signature and why we need this hack.

File: src/baseTokens.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  getTokenNameMulti,
+  getTokenSymbolMulti,
+} from "src/erc20";
+import { TokenTag } from "src/tags";
+import { getCurveTokenInfo } from "./curveToken";
+import { BaseTokenInfo, SimpleTokenInfo } from "./types";
+
+export const provider = hre.ethers.provider;
+
+export async function getBaseTokenInfos(
```
Comment (sentilesdal) [inline src/baseTokens.ts:15]: comment that this either returns a 'native' token or a curve lp token.  not sure if native is the right word there, but some quick explanation as to what's going on here :)

File: src/curveToken.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+import { CurveLpToken } from "./types";
+import { ethers } from "ethers";
+
+export const provider = hre.ethers.provider;
+const etherscanProvider = new ethers.providers.EtherscanProvider(
+  1,
+  process.env.ETHERSCAN_API_KEY
+);
+
+async function getGauges(): Promise<
```
Comment (sentilesdal) [inline src/curveToken.ts:13]: comment on what a Gauge is.  I'm still not clear after reading through curve's GH repo haha

File: src/curveToken.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+  } = await axios.get("https://api.curve.fi/api/getFactoryV2Pools");
+
+  if (!success) {
+    return;
+  }
+
+  return poolData;
+}
+
+async function getCurvePoolFromTokenAddress(address: string): Promise<string> {
```
Comment (sentilesdal) [inline src/curveToken.ts:96]: Comment on the strategies to get the pool address from the token address.

File: src/rootTokens.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
+import { ERC20__factory } from "elf-contracts-typechain/dist/types/factories/ERC20__factory";
+import hre from "hardhat";
+import { getCurveTokenInfo } from "./curveToken";
+import { TokenTag } from "./tags";
+import { CurveBaseToken, RootTokenInfo, SimpleRootToken } from "./types";
+
+export const provider = hre.ethers.provider;
+
+const ETH_CONSTANT = "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE";
```
[the remark this answers is not in the document]
Comment (Padraic-O-Mhuiris) [reply src/rootTokens.ts:9]: https://curve.readthedocs.io/exchange-cross-asset-swaps.html?highlight=eeee#initiating-a-swap seems to be according to this

File: src/baseTokens.ts (hunk 2 above)
Comment (Padraic-O-Mhuiris) [reply src/baseTokens.ts:15]: Done, I've use simpleTokenInfo as per the type

File: src/getTokenList.ts [hunk 6]
```diff
@@ -1,33 +1,42 @@
...
+    description: "The underlying base asset of a principal or yield token",
+  },
+  [TokenTag.ROOT]: {
+    name: "Root token",
+    description:
+      "This token is a member of a given LP token's pool assets. The LP token for that pool must also be a base token",
+  },
+  [TokenTag.CURVE]: {
+    name: "Curve LP token",
+    description: "A curve LP token",
```
Comment (DannyDelott) [inline src/getTokenList.ts:39]: Ohh I really like the `CURVE` tag for base assets that are LP tokens. That's a better model!

And I think we can get rid of `BASE` and `ROOT` tags entirely now. For tokens like dai, usdc, lusd, etc.. we don't need any tags or extensions on them. This basically treats them as boring erc20s, which is exactly what they are and token infos for `eP`, `eY`, `CURVE`, `CCPOOL`, etc.. will just reference them in their extensions.

Comment (Padraic-O-Mhuiris) [reply src/getTokenList.ts:39]: Hmm, I think we actually do need them, specifically in the case of `3crv` cos it's an LP token that is a root token, as in dependent on a base token like `MIM-3LP`

**Pull request closed without merging**